README.md
Audio recorder from microphone to a given file path.
No external dependencies, MediaRecorder is used for Android an AVAudioRecorder for iOS.

## Options
- bit rate (be careful with this one on iOs)
- sampling rate
- encoder

## Platforms

### Android
```xml
<uses-permission android:name="android.permission.RECORD_AUDIO" />
<!-- Optional, you'll have to check this permission by yourself. -->
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```
min SDK: 16 (29 if you use OPUS)

### iOs
```xml
<key>NSMicrophoneUsageDescription</key>
<string>We need to access to the microphone to record audio file</string>
```
min SDK: 8.0 (11 if you use OPUS)

## Supported encoders
```dart

enum AudioEncoder {
  /// Will output to MPEG_4 format container
  AAC,

  /// Will output to MPEG_4 format container
  AAC_LD,

  /// Will output to MPEG_4 format container
  AAC_HE,

  /// sampling rate should be set to 8kHz
  /// Will output to 3GP format container on Android
  AMR_NB,

  /// sampling rate should be set to 16kHz
  /// Will output to 3GP format container on Android
  AMR_WB,

  /// Will output to MPEG_4 format container
  /// /!\ SDK 29 on Android /!\
  /// /!\ SDK 11 on iOs /!\
  OPUS,
}
```

### Android
https://developer.android.com/reference/android/media/MediaRecorder.AudioEncoder
### iOs
https://developer.apple.com/documentation/coreaudiotypes/coreaudiotype_constants/1572096-audio_data_format_identifiers

## Usage
```dart
// Import package
import 'package:record/record.dart';

// Check and request permission
bool result = await Record.hasPermission();

// Start recording
await Record.start(
  path: 'aFullPath/myFile.m4a', // required
  encoder: AudioEncoder.AAC, // by default
  bitRate: 128000, // by default
  sampleRate: 44100, // by default
);

// Stop recording
await Record.stop();

// Get the state of the recorder
bool isRecording = await Record.isRecording();

// There's nothing to dispose, this done internally each time you call stop method.
// The plugin is aware of activity lifecycle.
// So exiting, your app or activity will stop the recording (but won't delete the 
// output file).
```

## Warnings
Be sure to check supported values from the given links above.

## Known issues
None.

CHANGELOG.md
## 2.1.1
* fix: Android warning "mediarecorder went away with unhandled events".

## 2.1.0+1
* Relaxing dart/flutter constraints for pub.dev analysis.

## 2.1.0
* Add pause/resume/isPaused features.
* Updated example to include pause.

## 2.0.0
* Add null safety support (based on v1.0.3).
* Updated example. Moved from 'audioplayers' to 'just_audio'.

## 1.0.3
* Allow recording from bluetooth device on iOS.

## 1.0.2
* Fix (for good) potential exception when closing beofre recorder is actually started (Android).

## 1.0.1
* Fix potential exception when closing beofre recorder is actually started (Android).
* Fix potential NPE on permission result (Android).
* Updade dependencies.

## 1.0.0

* Flutter 1.12.0 is now minimum version (Flutter plugin API v2).
* Finalize Record API.
* Request permission on Android too.
* Automatically stop recording when leaving activity / app.
* Add all codecs with cross platform compatibility.

## 0.2.1

* Fix broken build on Android.

## 0.2.0

* Fix broken build on Android.

## 0.1.0

* Initial release.
* Implementations for Android and iOs.
* Example.


lib/record.dart
import 'dart:async';

import 'package:flutter/services.dart';

/// Record class bridging to native recorders given by platform SDKs.
///
/// The plugin is aware of activity lifecycle.
/// So exiting, your app or activity will stop the recording (but won't delete the
/// output file).
class Record {
  static const MethodChannel _channel = const MethodChannel(
    'com.llfbandit.record',
  );

  /// Starts new recording session.
  ///
  /// [path]: The output path file. Required.
  /// [encoder]: The audio encoder to be used for recording.
  /// [bitRate]: The audio encoding bit rate in bits per second.
  /// [samplingRate]: The sampling rate for audio in samples per second.
  static Future<void> start({
    required String path,
    AudioEncoder encoder = AudioEncoder.AAC,
    int bitRate = 128000,
    double samplingRate = 44100.0,
  }) {
    return _channel.invokeMethod('start', {
      "path": path,
      "encoder": encoder.index,
      "bitRate": bitRate,
      "samplingRate": samplingRate,
    });
  }

  /// Stops recording session and release internal recorder resource.
  static Future<void> stop() {
    return _channel.invokeMethod('stop');
  }

  /// Pauses recording session.
  ///
  /// Note: Usable on Android API >= 24(Nougat). Does nothing otherwise.
  static Future<void> pause() {
    return _channel.invokeMethod('pause');
  }

  /// Resumes recording session after [pause].
  ///
  /// Note: Usable on Android API >= 24(Nougat). Does nothing otherwise.
  static Future<void> resume() {
    return _channel.invokeMethod('resume');
  }

  /// Checks if there's valid recording session.
  /// So if session is paused, this method will still return [true].
  static Future<bool> isRecording() async {
    final result = await _channel.invokeMethod<bool>('isRecording');
    return result ?? false;
  }

  /// Checks if recording session is paused.
  static Future<bool> isPaused() async {
    final result = await _channel.invokeMethod<bool>('isPaused');
    return result ?? false;
  }

  /// Checks and requests for audio record permission.
  static Future<bool> hasPermission() async {
    final result = await _channel.invokeMethod<bool>('hasPermission');
    return result ?? false;
  }
}

/// Audio encoder to be used for recording.
enum AudioEncoder {
  /// Will output to MPEG_4 format container.
  AAC,

  /// Will output to MPEG_4 format container.
  AAC_LD,

  /// Will output to MPEG_4 format container.
  AAC_HE,

  /// sampling rate should be set to 8kHz.
  /// Will output to 3GP format container on Android.
  AMR_NB,

  /// sampling rate should be set to 16kHz.
  /// Will output to 3GP format container on Android.
  AMR_WB,

  /// Will output to MPEG_4 format container.
  /// /!\ SDK 29 on Android /!\
  /// /!\ SDK 11 on iOs /!\
  OPUS,
}
